### PR TITLE
Fix panic when comparing snapshots

### LINKED ISSUE #34: Cluster config modification events are not generated
### Describe the bug

Cluster config modification events are sometimes not appearing as Kubernetes events even though the operator code clearly is executed. This happens infrequently when running tests against either a DIND cluster or dev, but never when the failing tests are run in isolation.

This perhaps suggests that some events get overwritten or dropped. This is surprising given that the `k8s.io/client-go/tools/record/EventRecorder` can queue up to 1000 events and logs an error when the same even has reached the max retry count.

Failing test: Allowable cluster modifications > cluster config file changes
```
[10] Allowable cluster modifications
[10] /home/sebbonnet/gowork/src/github.com/sky-uk/core-operators/cassandra-operator/test/e2e/parallel/modification/modification_test.go:50
[10]   cluster config file changes
[10]   /home/sebbonnet/gowork/src/github.com/sky-uk/core-operators/cassandra-operator/test/e2e/parallel/modification/modification_test.go:175
[10]     should trigger a rolling restarts of the cluster stateful set when the custom config file is deleted [It]
[10]     /home/sebbonnet/gowork/src/github.com/sky-uk/core-operators/cassandra-operator/test/e2e/parallel/modification/modification_test.go:249
[10]
[10]     Timed out after 30.001s.
[10]     Expected event {Normal ClusterUpdate Custom config deleted for cluster test-cassandra-operator.mycluster-fdevi 2019-03-22 18:03:50.090478874 +0000 GMT m=+251.420548413}. Actual event: [{{ } {mycluster-fdevi.158e59c9b7eb31e1  test-cassandra-operator /api/v1/namespaces/test-
cassandra-operator/events/mycluster-fdevi.158e59c9b7eb31e1 69d314af-4ccc-11e9-88c4-0624fb7941ea 507389776 0 2019-03-22 18:00:44 +0000 GMT <nil> <nil> map[] map[] [] nil [] } 
[...]
[10]     /home/sebbonnet/gowork/src/github.com/sky-uk/core-operators/cassandra-operator/test/e2e/parallel/modification/modification_test.go:265
```
 ### Expected Behavior
The tests should consistently pass regardless of whether they are run in parallel or not.

 ### Steps to Reproduce
1. `cd $GOPATH/src/github.com/sky-uk/cassandra-operator/`
2. `make install-all`
3. `cd cassandra-operator`
4. `E2E_TEST=test/e2e/parallel/modification make e2e-test`

 ### Your Environment
* Kubernetes version: `1.10.12`
* OS: Fedora 28, 12 cores
* Cassandra operator docker image: `skyuk/cassandra-operator:v0.70.0-alpha`
* Cassandra cluster definition: those created by the tests for this git revision `v0.70.0-alpha`

### PR DESCRIPTION
**Pull request by sebbonnet**
Fix #34 

When a cluster definition was modified, there was a scenario that would cause a panic resulting in the update to be missed. The panic would occur when comparing the snapshots properties between the old cluster definition and the new one when one of the snapshot had a nil property.
This PR fixes the panic and add tests to verify the snapshot comparison.

Also changes the delete propagation policy to Background when cleaning up `cassandra` resource.
The previous value of `Foreground` (`Orphan` too) was causing an additional (and unneeded) cluster update event during the test cleanup which in some cases would cause the panic we've seen in #34 and so affect other tests that could be running in parallel. Note that this problem is more visible on machines with a high number of cores. With 2 cores, we've actually never seen this issue on our travis environment, whereas I can almost always see it on my 12 core machines.
I feel `Background` is actually a better fit, given that we delete all other resources upfront and as `kubectl` default policy behaves in the same way (at least for the `cassandra` resource)

### CODE REVIEW

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		})
+
+		It("should be found different when cleanup timeout have differnt values", func() {
+			snapshot1.RetentionPolicy.CleanupTimeoutSeconds = ptr.Int32(30)
+			Expect(SnapshotCleanupPropertiesUpdated(snapshot1, snapshot2)).To(BeTrue())
+			Expect(SnapshotCleanupPropertiesUpdated(snapshot2, snapshot1)).To(BeTrue())
+		})
+
+	})
+})
```
Comment (wallrj) [inline cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go]: :grey_question: Nice tests for the API helpers. We should look into adding code coverage reporting. https://github.com/sky-uk/cassandra-operator/issues/124

File: cassandra-operator/pkg/operator/operations/receiver.go [hunk 2]
```diff
@@ -64,6 +65,7 @@ func NewEventReceiver(clusters map[string]*cluster.Cluster, clusterAccessor *clu
 
 // Receive receives operator events and delegates their processing to the appropriate handler
 func (r *Receiver) Receive(event *dispatcher.Event) {
+	log.Debugf("Event type %s received for key %s", event.Kind, event.Key)
```
Comment (wallrj) [inline cassandra-operator/pkg/operator/operations/receiver.go]: :grey_question: You could use logrus fields here and a local logger where the event kind and key are added to every log message in this method call.

File: cassandra-operator/pkg/operator/operations/update_cluster.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ func (o *UpdateClusterOperation) Execute() {
 
 	clusterChanges, err := o.adjuster.ChangesForCluster(oldCluster, newCluster)
 	if err != nil {
+		log.Warnf("unable to generate patch for cluster %s.%s: %v", newCluster.Namespace, newCluster.Name, err)
```
Comment (wallrj) [inline cassandra-operator/pkg/operator/operations/update_cluster.go]: :grey_question: Could use logrus fields and `.WithError` here, if you feel like it.

File: cassandra-operator/go.mod [hunk 4]
```diff
@@ -5,6 +5,7 @@ go 1.12
 require (
 	github.com/PaesslerAG/gval v0.1.1 // indirect
 	github.com/PaesslerAG/jsonpath v0.1.0
+	github.com/davecgh/go-spew v1.1.1
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/go.mod]: ah right, some of our transitive dependencies must be using `pretty`. I had a look at it, but I preferred the output of `spew`

File: cassandra-operator/cmd/operator/main.go [hunk 5]
```diff
@@ -68,7 +68,7 @@ func startOperator(_ *cobra.Command, _ []string) error {
 		MetricPollInterval:    metricPollInterval,
 		AllowEmptyDir:         allowEmptyDir,
 	}
-	log.Infof("Starting Cassandra operator with config: %v", operatorConfig)
+	log.Infof("Starting Cassandra operator with config: %+v", operatorConfig)
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/main.go]: indeed - I've commented on that issue

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra.go [hunk 6]
```diff
@@ -64,3 +66,23 @@ func GetDatacenter(c *v1alpha1.Cassandra) string {
...
+		!reflect.DeepEqual(snapshot1.Keyspaces, snapshot2.Keyspaces)
+}
+
+// SnapshotCleanupPropertiesUpdated returns false snapshot1 and snapshot2 have the same retention policy regardless of whether it is enabled or not
+func SnapshotCleanupPropertiesUpdated(snapshot1 *v1alpha1.Snapshot, snapshot2 *v1alpha1.Snapshot) bool {
+	return snapshot1.RetentionPolicy != nil && snapshot2.RetentionPolicy != nil &&
+		(snapshot1.RetentionPolicy.CleanupSchedule != snapshot2.RetentionPolicy.CleanupSchedule ||
+			!reflect.DeepEqual(snapshot1.RetentionPolicy.CleanupTimeoutSeconds, snapshot2.RetentionPolicy.CleanupTimeoutSeconds) ||
+			!reflect.DeepEqual(snapshot1.RetentionPolicy.RetentionPeriodDays, snapshot2.RetentionPolicy.RetentionPeriodDays))
+}
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra.go]: Let's do that in a different PR as I also think the default should be Enabled=true not false as it is now.
See https://github.com/sky-uk/cassandra-operator/issues/72#issuecomment-501621943

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	"github.com/sky-uk/cassandra-operator/cassandra-operator/pkg/apis/cassandra/v1alpha1"
+	"github.com/sky-uk/cassandra-operator/cassandra-operator/pkg/util/ptr"
+	"github.com/sky-uk/cassandra-operator/cassandra-operator/test"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestDispatcher(t *testing.T) {
```
Comment (alangibson01) [inline cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go]:
```suggestion
func TestHelpers(t *testing.T) {
```

**Comment by sebbonnet** [comment]
@wallrj It does feel like we're doing the defaulting too late.
We currently handle default values when creating the `Cluster` struct, which happens after we identify the list of operations to apply based on the `Cassandra` struct (these would have non-defaulted pointers). Perhaps we should move the defaulting out of the `Cluster` struct and apply it right after the Cassandra resource is unmarshalled?
I think we would need to extract the defaulting anyway once we implement the mutating webhook, so that would be a better time to review this issue

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go (hunk 7 above)
Comment (sebbonnet) [reply cassandra-operator/pkg/apis/cassandra/v1alpha1/helpers/cassandra_test.go]: haha - old good copy and paste :)

**Comment by wallrj** [comment]
Yeah, that's what I was getting at in the standup. 
It seems like there ought to be a default for `TimeoutSeconds` `CleanupTimeoutSeconds`, and `RetentionPeriodDays` at least. 
And we could apply the defaults in the informer callback functions (or eventually in the work queue worker function) immediately after the resource is unmarshalled, then we need never worry about these "leaf" fields being nil in our helper functions.

**Comment by wallrj** [comment]
@sebbonnet Regarding the changes to the Deletion policy...I've realised what's bothering me about it...it's occurred to me that we're not handling the "tombstone" situation in the informer delete callback.
The delete callback doesn't always get a `Cassandra` object. Instead it gets a `cache.DeletedFinalStateUnknown` object.
I wonder if that might be the cause of the panics you're seeing.
See:
 * https://github.com/kubernetes/sample-controller/blob/master/controller.go#L356
 * and `DeletionHandlingMetaNamespaceKeyFunc`, https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L257

...which is probably what we'll use when we switch to the work queue architecture.

**Comment by sebbonnet** [comment]
@wallrj that's interesting, but this is not the scenario I have observed.
The panic occurs only when an UPDATE_CLUSTER is detected and for a fairly edge case: both old and new Cassandra have to have a Snapshot with the same Schedule and a nil Timeout. If Schedule are different panic doesn't occur as the comparison returns right away.

Now the interesting thing is that this UPDATE_CLUSTER was triggered when the Cassandra definition was deleted with a Foreground policy. Deleting with a Background policy doesn't trigger an UPDATE_CLUSTER, but only a DELETE_CLUSTER. On the other hand deleting with a Foreground or Orphan policy triggers an UPDATE_CLUSTER following by a DELETE_CLUSTER.
I've added additional log statement to capture the object deleted, but I couldn't see any evidence of the tombstone you mentioned whatever the propagation policy I used.

Here is some logs I've captured that illustrate that:

**Default policy (Background for Cassandra definition)**
```
> kubectl --context dev -n cassandra-operator-stubbed-functional delete cassandra mycluster-glx2x
cassandra.core.sky.uk "mycluster-glx2x" deleted 

> (operator logs)

time="2019-06-13T15:52:25Z" level=debug msg="Cluster delete detected obj: <*>(0xc0002dc000){TypeMeta:&TypeMeta{Kind:,APIVersion:,} ObjectMeta:&ObjectMeta{Name:mycluster-glx2x,GenerateName:,Namespace:cassandra-operator-stubbed-functional,SelfLink:/apis/core.sky.uk/v1alpha1/namespaces/cassandra-operator-stubbed-functional/cassandras/mycluster-glx2x,UID:fccfd116-8df2-11e9-ad6a-0624fb7941ea,ResourceVersion:602648737,Generation:1,CreationTimestamp:2019-06-13 15:50:37 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],} Spec:{Datacenter:<nil> Racks:[{Name:a Zone:eu-west-1a StorageClass:standard-zone-a Replicas:1}] UseEmptyDir:<*>(0xc00063bbbf)true Pod:{BootstrapperImage:<*>(0xc0003bf120)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-bootstrapper:vaf2865d SidecarImage:<*>(0xc0003bf160)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-sidecar:vaf2865d Image:<*>(0xc0003bf140)registry.tools.cosmic.sky/core-engineering/core-operators-test/fake-cassandra:vaf2865d StorageSize:0 Memory:1Gi CPU:0 LivenessProbe:<*>(0xc00025a1b0){FailureThreshold:<*>(0xc00063bb40)3 InitialDelaySeconds:<*>(0xc00063bb44)1 PeriodSeconds:<*>(0xc00063bb48)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>} ReadinessProbe:<*>(0xc00025a1e0){FailureThreshold:<*>(0xc00063bb58)3 InitialDelaySeconds:<*>(0xc00063bb5c)1 PeriodSeconds:<*>(0xc00063bb60)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>}} Snapshot:<*>(0xc000688f80){Image:<*>(0xc0003bf180)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-snapshot:vaf2865d Schedule:59 23 * * * Keyspaces:[keyspace1 keyspace3] TimeoutSeconds:<nil> RetentionPolicy:<*>(0xc00025a210){Enabled:true RetentionPeriodDays:<*>(0xc00063bbb0)10 CleanupSchedule:11 22 1 * * CleanupTimeoutSeconds:<*>(0xc00063bb9c)5}}} Status:{}}"                 
time="2019-06-13T15:52:25Z" level=debug msg="Event received" key=cassandra-operator-stubbed-functional.mycluster-glx2x type=DELETE_CLUSTER
time="2019-06-13T15:52:25Z" level=info msg="Event will trigger 3 operations" key=cassandra-operator-stubbed-functional.mycluster-glx2x type=DELETE_CLUSTER                                                                                                                               
time="2019-06-13T15:52:25Z" level=debug msg="Executing operation delete cluster cassandra-operator-stubbed-functional.mycluster-glx2x" key=cassandra-operator-stubbed-functional.mycluster-glx2x type=DELETE_CLUSTER                                                                     
time="2019-06-13T15:52:25Z" level=info msg="Cassandra cluster definition deleted for cluster: cassandra-operator-stubbed-functional.mycluster-glx2x"                                                                                                                                     
time="2019-06-13T15:52:25Z" level=info msg="Removing node cluster:cassandra-operator-stubbed-functional.mycluster-glx2x, pod:mycluster-glx2x-a-0, rack: from metrics"                                                                                                                    
time="2019-06-13T15:52:25Z" level=debug msg="Executing operation delete snapshot schedule for cluster cassandra-operator-stubbed-functional.mycluster-glx2x" key=cassandra-operator-stubbed-functional.mycluster-glx2x type=DELETE_CLUSTER                                               
time="2019-06-13T15:52:25Z" level=debug msg="Executing operation delete snapshot cleanup schedule for cluster cassandra-operator-stubbed-functional.mycluster-glx2x" key=cassandra-operator-stubbed-functional.mycluster-glx2x type=DELETE_CLUSTER
```

**Orphan policy**

I coudn't find a way to use a Foreground policy with `kubeclt`, but doing it programmatically via the casandra clientset api triggers the same behaviour for Orphans and Foreground policy as far as the Cassandra events are concerned.  
If you look at the output closely for the UPDATE_CLUSTER event, you will see that the `SuccessThreshold` and `TimeoutSeconds` for both probes have changed. I have no idea why.

```
> kubectl --context dev -n cassandra-operator-stubbed-functional delete cassandra mycluster-h0jp2 --cascade=false

> (operator logs)
time="2019-06-13T15:47:54Z" level=debug msg="Cluster update detected for cassandra-operator-stubbed-functional.mycluster-h0jp2, old: {Datacenter:<nil> Racks:[{Name:a Zone:eu-west-1a StorageClass:standard-zone-a Replicas:1}] UseEmptyDir:<*>(0xc00083c67f)true Pod:{BootstrapperImage:<*>(0xc000091c00)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-bootstrapper:vaf2865d SidecarImage:<*>(0xc000091c40)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-sidecar:vaf2865d Image:<*>(0xc000091c20)registry.tools.cosmic.sky/core-engineering/core-operators-test/fake-cassandra:vaf2865d StorageSize:0 Memory:1Gi CPU:0 LivenessProbe:<*>(0xc000020780){FailureThreshold:<*>(0xc00083c600)3 InitialDelaySeconds:<*>(0xc00083c604)1 PeriodSeconds:<*>(0xc00083c608)1 SuccessThreshold:<*>(0xc00004933c)1 TimeoutSeconds:<*>(0xc000049340)5} ReadinessProbe:<*>(0xc000020810){FailureThreshold:<*>(0xc00083c618)3 InitialDelaySeconds:<*>(0xc00083c61c)1 PeriodSeconds:<*>(0xc00083c620)1 SuccessThreshold:<*>(0xc000049350)1 TimeoutSeconds:<*>(0xc000049354)5}} Snapshot:<*>(0xc000228f40){Image:<*>(0xc000091c50)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-snapshot:vaf2865d Schedule:59 23 * * * Keyspaces:[keyspace1 keyspace3] TimeoutSeconds:<nil> RetentionPolicy:<*>(0xc000020840){Enabled:true RetentionPeriodDays:<*>(0xc00083c670)10 CleanupSchedule:11 22 1 * * CleanupTimeoutSeconds:<*>(0xc00083c65c)5}}} \nnew: {Datacenter:<nil> Racks:[{Name:a Zone:eu-west-1a StorageClass:standard-zone-a Replicas:1}] UseEmptyDir:<*>(0xc0008f696f)true Pod:{BootstrapperImage:<*>(0xc0007a0e20)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-bootstrapper:vaf2865d SidecarImage:<*>(0xc0007a0e50)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-sidecar:vaf2865d Image:<*>(0xc0007a0e40)registry.tools.cosmic.sky/core-engineering/core-operators-test/fake-cassandra:vaf2865d StorageSize:0 Memory:1Gi
CPU:0 LivenessProbe:<*>(0xc0004837d0){FailureThreshold:<*>(0xc0008f68f0)3 InitialDelaySeconds:<*>(0xc0008f68f4)1 PeriodSeconds:<*>(0xc0008f68f8)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>} ReadinessProbe:<*>(0xc000483800){FailureThreshold:<*>(0xc0008f6908)3 InitialDelaySeconds:<*>(0xc0008f690c)1 PeriodSeconds:<*>(0xc0008f6910)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>}} Snapshot:<*>(0xc00086a900){Image:<*>(0xc0007a0e60)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-snapshot:vaf2865d Schedule:59 23 * * * Keyspaces:[keyspace1 keyspace3] TimeoutSeconds:<nil> RetentionPolicy:<*>(0xc000483830){Enabled:true RetentionPeriodDays:<*>(0xc0008f6960)10 CleanupSchedule:11 22 1 * * CleanupTimeoutSeconds:<*>(0xc0008f694c)5}}}"                                                                                             
time="2019-06-13T15:47:54Z" level=debug msg="Cluster delete detected obj: <*>(0xc000189440){TypeMeta:&TypeMeta{Kind:,APIVersion:,} ObjectMeta:&ObjectMeta{Name:mycluster-h0jp2,GenerateName:,Namespace:cassandra-operator-stubbed-functional,SelfLink:/apis/core.sky.uk/v1alpha1/namespaces/cassandra-operator-stubbed-functional/cassandras/mycluster-h0jp2,UID:635264d4-8df2-11e9-8f55-0a7a8c8bc22a,ResourceVersion:602645587,Generation:2,CreationTimestamp:2019-06-13 15:46:19 +0000 UTC,DeletionTimestamp:2019-06-13 15:47:54 +0000 UTC,DeletionGracePeriodSeconds:*0,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[orphan],ClusterName:,Initializers:nil,ManagedFields:[],} Spec:{Datacenter:<nil> Racks:[{Name:a Zone:eu-west-1a StorageClass:standard-zone-a Replicas:1}] UseEmptyDir:<*>(0xc0008f774f)true Pod:{BootstrapperImage:<*>(0xc0007a1080)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-bootstrapper:vaf2865d SidecarImage:<*>(0xc0007a10b0)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-sidecar:vaf2865d Image:<*>(0xc0007a1090)registry.tools.cosmic.sky/core-engineering/core-operators-test/fake-cassandra:vaf2865d StorageSize:0 Memory:1Gi CPU:0 LivenessProbe:<*>(0xc000483aa0){FailureThreshold:<*>(0xc0008f76d0)3 InitialDelaySeconds:<*>(0xc0008f76d4)1 PeriodSeconds:<*>(0xc0008f76d8)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>} ReadinessProbe:<*>(0xc000483ad0){FailureThreshold:<*>(0xc0008f76e8)3 InitialDelaySeconds:<*>(0xc0008f76ec)1 PeriodSeconds:<*>(0xc0008f76f0)1 SuccessThreshold:<nil> TimeoutSeconds:<nil>}} Snapshot:<*>(0xc00086ab80){Image:<*>(0xc0007a10c0)registry.tools.cosmic.sky/core-engineering/core-operators-test/cassandra-snapshot:vaf2865d Schedule:59 23 * * * Keyspaces:[keyspace1 keyspace3] TimeoutSeconds:<nil> RetentionPolicy:<*>(0xc000483b00){Enabled:true RetentionPeriodDays:<*>(0xc0008f7740)10 CleanupSchedule:11 22 1 * * CleanupTimeoutSeconds:<*>(0xc0008f772c)5}}}
Status:{}}"
time="2019-06-13T15:47:54Z" level=debug msg="Event received" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=UPDATE_CLUSTER
time="2019-06-13T15:47:54Z" level=info msg="Event will trigger 1 operations" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=UPDATE_CLUSTER                                                                                                                               
time="2019-06-13T15:47:54Z" level=debug msg="Executing operation update cluster cassandra-operator-stubbed-functional.mycluster-h0jp2" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=UPDATE_CLUSTER                                                                     
time="2019-06-13T15:47:54Z" level=info msg="Cluster definition has been updated for cluster cassandra-operator-stubbed-functional.mycluster-h0jp2"                                                                                                                                       
time="2019-06-13T15:47:54Z" level=debug msg="Event received" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=DELETE_CLUSTER
time="2019-06-13T15:47:54Z" level=info msg="Event will trigger 3 operations" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=DELETE_CLUSTER                                                                                                                               
time="2019-06-13T15:47:54Z" level=debug msg="Executing operation delete cluster cassandra-operator-stubbed-functional.mycluster-h0jp2" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=DELETE_CLUSTER                                                                     
time="2019-06-13T15:47:54Z" level=info msg="Cassandra cluster definition deleted for cluster: cassandra-operator-stubbed-functional.mycluster-h0jp2"                                                                                                                                     
time="2019-06-13T15:47:54Z" level=info msg="Removing node cluster:cassandra-operator-stubbed-functional.mycluster-h0jp2, pod:mycluster-h0jp2-a-0, rack: from metrics"                                                                                                                    
time="2019-06-13T15:47:54Z" level=debug msg="Executing operation delete snapshot schedule for cluster cassandra-operator-stubbed-functional.mycluster-h0jp2" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=DELETE_CLUSTER                                               
time="2019-06-13T15:47:54Z" level=debug msg="Executing operation delete snapshot cleanup schedule for cluster cassandra-operator-stubbed-functional.mycluster-h0jp2" key=cassandra-operator-stubbed-functional.mycluster-h0jp2 type=DELETE_CLUSTER       
```

**Comment by sebbonnet** [comment]
@wallrj  regarding this:
> If you look at the output closely for the UPDATE_CLUSTER event, you will see that the SuccessThreshold and TimeoutSeconds for both probes have changed. I have no idea why.

I digged a little further and can confirm your suspicion: the changes in the probes properties is because we update the Cassandra `cached` object when initially when defaulting values for the probes.
When the `Cassandra` object is deleted via either a Orphan or Foreground deletion policy, the informer performs an update (possibly because the Cassandra object is marked as being deleted for Foreground or its link with children remove for Orphan) which is then processed by our `receiver` and proceeded with the update operation as the old and new Cassandra object differ.
If I change our code to do a DeepCopy of the Cassandra `cached` object, the informed still performs an update, but this time our `receiver` sees no difference and proceeds no further.

**Pull request merged**